src/Aspire.Dashboard/Components/Controls/ApplicationName.razor

```diff
@@ -0,0 +1,24 @@
+﻿@namespace Aspire.Dashboard.Components
+@using Aspire.Dashboard.Model
+@using System.Globalization
+@inject IDashboardClient DashboardClient
+
+@code {
+    [Parameter]
+    public required string ResourceName { get; init; }
+
+    [Parameter]
+    public required IStringLocalizer Loc { get; init; }
+
+    private string? applicationName;
+
+    protected override async Task OnInitializedAsync()
+    {
+        // We won't have an application name until the client has connected to the server.
+        await DashboardClient.WhenConnected;
```

Comment:
I think this doesn't create a leak. e.g. visiting lots of pages while there is no connection will GC the waiting state machine when the page is GCed. Does that line up with your understanding?

It could be made explicit by creating a CTS in the control, [using the token when awaiting `WhenConnected`](https://github.com/dotnet/runtime/issues/14991#issuecomment-478072333), and then canceling the CTS in component's dispose.

Reply:
Actually, this will leak &mdash; good spot. The `WhenConnected` task must retain a collection of continuations, rooting the caller's state machine. That machine captures `this`, meaning the whole component would be retained.

I was concerned component disposal might not occur until after `OnInitializedAsync` completed, but the [docs say](https://learn.microsoft.com/en-us/aspnet/core/blazor/components/lifecycle?view=aspnetcore-8.0):

> Disposal can occur at any time, including during [component initialization](https://learn.microsoft.com/en-us/aspnet/core/blazor/components/lifecycle?view=aspnetcore-8.0#component-initialization-oninitializedasync).

It seems that handling dispose in the way you describe would be safest.

> [using the token when awaiting `WhenConnected`](https://github.com/dotnet/runtime/issues/14991#issuecomment-478072333)

I think the built-in [`Task.WaitAsync`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.waitasync?view=net-8.0) is a better option these days.

Reply:
As an aside, I couldn't quickly work out how to define `DisposeAsync` in the `.razor` file, so introduced a code behind file. If you know a way to do that I'd be interested to hear it.

Reply:
> Actually, this will leak — good spot. The `WhenConnected` task must retain a collection of continuations, rooting the caller's state machine. That machine captures `this`, meaning the whole component would be retained.

Did you test on debug? I'm fairly sure the behavior differs between debug compilation and release compilation.

Either way, I like being explicit and prefer having the CTS.

Reply:
My assumption is that the `await` ultimately, via some state machine helpers, creates a delegate for the following code, and passes it to `ContinueWith`, which queues it, and also roots it. I think the .NET debugger can stash task chain links on tasks directly, to power debugging tools.